pkg/kp/rollstore/consul_store.go

```diff
@@ -0,0 +1,555 @@
+		LeaveOld:        leaveOld,
+		RollDelay:       rollDelay,
+	}
+
+	return s.attemptRUCreation(u, renewalErrCh)
+}
+
+// Creates a rolling update that may or may not already have an existing old
+// RC. If one matches the oldRCSelector, it will be used as the old RC in the
+// new update.  If one does not exist, a "dummy" old RC will be created that is
+// identical to the specifications for the new RC.
+// Returns an error if the old RC exists but is part of another RU, or if
+// the label selector returns more than one match.
+func (s consulStore) CreateRollingUpdateFromOneMaybeExistingWithLabelSelector(
+	oldRCSelector klabels.Selector,
+	desiredReplicas int,
+	minimumReplicas int,
+	leaveOld bool,
+	rollDelay time.Duration,
+	newRCManifest pods.Manifest,
+	newRCNodeSelector klabels.Selector,
+	newRCPodLabels klabels.Set,
+) (u roll_fields.Update, err error) {
+	// This function may or may not create old and new RCs and subsequently
+	// fail, so we defer a function that does any cleanup (if applicable)
+	var cleanupOldRC func()
+	var cleanupNewRC func()
+
+	defer func() {
+		if err != nil {
+			if cleanupOldRC != nil {
+				cleanupOldRC()
+			}
+
+			if cleanupNewRC != nil {
+				cleanupNewRC()
+			}
+		}
+	}()
+
+	session, renewalErrCh, err := s.newRUCreationSession()
+	if err != nil {
+		return roll_fields.Update{}, err
+	}
+
+	// Check if any RCs match the oldRCSelector
+	matches, err := s.labeler.GetMatches(oldRCSelector, labels.RC)
+	if err != nil {
+		return roll_fields.Update{}, err
+	}
+
+	if len(matches) > 1 {
+		return roll_fields.Update{}, util.Errorf(
+			"Can't create update: old RC selector %s was ambiguous, provided %d matches",
+			oldRCSelector.String(),
+			len(matches),
+		)
+	}
+
+	var oldRCID rc_fields.ID
+	if len(matches) == 1 {
+		oldRCID = rc_fields.ID(matches[0].ID)
+	} else {
+		if leaveOld {
+			return roll_fields.Update{}, util.Errorf(
+				"Can't create an update with LeaveOld set if there is no old RC (sel=%s)",
+				oldRCSelector.String(),
+			)
+		}
+
+		// Create the old RC using the same info as the new RC, it'll be
+		// removed when the update completes anyway
+		rc, err := s.rcstore.Create(newRCManifest, newRCNodeSelector, newRCPodLabels)
+		if err != nil {
+			return roll_fields.Update{}, err
+		}
+
+		oldRCID = rc.ID
+		cleanupOldRC = func() {
+			err = s.rcstore.Delete(oldRCID, false)
+			if err != nil {
+				s.logger.WithError(err).Errorf("Unable to cleanup newly-created old RC %s after update creation failure:", oldRCID)
+			}
+		}
+	}
+
+	// Lock the old RC to guarantee that no new updates can use it
+	err = s.lockRCs(rc_fields.IDs{oldRCID}, session)
+	if err != nil {
+		return roll_fields.Update{}, err
+	}
+
+	// Check for updates that exist that operate on the old RC
+	err = s.checkForConflictingUpdates(rc_fields.IDs{oldRCID})
+	if err != nil {
+		return roll_fields.Update{}, err
+	}
+
+	// Create the new RC
+	rc, err := s.rcstore.Create(newRCManifest, newRCNodeSelector, newRCPodLabels)
```

Comment:
In the previous function you attempted to drain the `sessionRenewalErr` channel before doing this, why not here?